Convert network files [2/n]

dopelsunce opened the pull request:
- This PR mainly finished converting functions and structs declared in `nameser.h` and `resolv.h`. The function definitions are scattered across a dozen `.c` files.
- Touched a few other headers (such as `errno.h`) to enable some functions or structs (such as `__errno_location`) to be used in checked scopes. 
- Replaced some nt_arary_ptrs with array_ptrs to simplify bounds declarations.
- Calls to `mem*` and `str*` are temporarily put in unchecked scopes, which can be removed once #2 and its follow-up PRs are merged.
- Tested on libc-test.

dopelsunce committed:
Rename occurrences of ptr.

dopelsunce committed:
nameser.h: add bounds safety interface

dopelsunce committed:
ns_parse.c: check a few functions

dopelsunce committed:
resolve.h: add bounds safety interfaces

dopelsunce committed:
ns_parse.c: remove checked scope for ns_skiprr

dopelsunce committed:
add comments and spacing to bounds-safe interfaces

dopelsunce committed:
add stdchecked.h

dopelsunce committed:
use longer aliases; remove header dependency

dopelsunce committed:
use nt_array_ptr where appropriate

dopelsunce committed:
update comments to utility functions in nameser.h

dopelsunce committed:
annotate function definitions as well

dopelsunce committed:
fix comments for ns_put functions

dopelsunce committed:
WIP: convert network files

dopelsunce committed:
Merge branch 'master' of https://github.com/microsoft/checkedc-musl into cconv

dopelsunce committed:
Make more functions checked in network directory

dopelsunce committed:
Fix typos and remove comments

dopelsunce committed:
Fix spacing

mgrang commented on src/locale/dcngettext.c:
```diff
@@ -81,7 +81,7 @@ char *bindtextdomain(const char *domainname, const char *dirname)
 
 	UNLOCK(lock);
-	
+
```
Remove extra newline.

dopelsunce replied:
Not a new line. It seems the spaces are removed. I think this is automatically done by my editor. Please let me know if I should restore the spaces.

mgrang replied:
Yes, please restore spaces. We want to make diff'ing the converted code with an unconverted one easier in future.

mgrang commented on src/network/dn_comp.c:
```diff
@@ -3,14 +3,26 @@
 		while (*s & 0xc0) {
 			if ((*s & 0xc0) != 0xc0) return 0;
-			s = base + ((s[0]&0x3f)<<8 | s[1]);
+			// At this point s must have length at least 1 (s[0] cannot be 0).
```
Why are we manually widening bounds here? The Checked C compiler can automatically widen bounds for null-terminated ptrs.

dopelsunce replied:
Thanks for pointing that out. I tried to remove the bounds cast and the compiler would emit an error on out-of-bounds access for `s[1]`. It seems currently the compiler cannot automatically widen the bounds for this particular piece of code (I used the current master branch). 

I tried removing the assignment to `s`, and got the same error:
```
while ((*s & 0xc0)) {
	if ((*s & 0xc0) != 0xc0) return 0;
	base + ((s[0]&0x3f)<<8 | s[1]); // rejected
}
```

I can get rid of the error after adding a _redundant_ if guard:
```
while ((*s & 0xc0)) {
	if ((*s & 0xc0) != 0xc0) return 0;
	if (*s) base + ((s[0]&0x3f)<<8 | s[1]); // compiled
}
```

But the code below is still rejected with the assignment back to `s`:
```
while ((*s & 0xc0)) {
	if ((*s & 0xc0) != 0xc0) return 0;
	if (*s) s = base + ((s[0]&0x3f)<<8 | s[1]); // rejected
}
```

__Question__: Will the analysis be improved in the future so the code snippet above can pass? __Would you suggest I keep the bounds cast for now?__ Thanks!

P.S. I updated the code to use `_Dynamic_bounds_cast` instead. There is no need for the unchecked block now.

mgrang commented on src/network/dn_expand.c:
```diff
@@ -1,28 +1,35 @@
-			if (j >= end-p || j >= dend-dest) return -1;
-			while (j--) *dest++ = *p++;
+			if (j >= end-p || j >= dend-dst) return -1;
+			while (j--) *dst++ = *p++;
```
Don't you need to break `*dst++` into `*dst` followed by `dst++` because of the runtime error that you reported?

dopelsunce replied:
`dst` does not use self-referential bounds. The undefined behavior in https://github.com/microsoft/checkedc-clang/issues/863 involves incrementing a pointer with a self-referential bound.

mgrang replied:
Can you add a clarifying comment to https://github.com/microsoft/checkedc-clang/issues/863 that the issue is only for pointers with self-referential bounds?

dopelsunce replied:
Added. Thanks!

dopelsunce committed:
Use dynamic_bounds_cast and restore a tab

kkjeer commented on src/network/dn_comp.c:
```diff
@@ -33,19 +45,26 @@ static int getlens(unsigned char *lens, const char *s, int l)
-static int match(int *offset, const unsigned char *base, const unsigned char *dn,
-	const char *end, const unsigned char *lens, int nlen)
+static int match(_Ptr<int> offset,
+	_Nt_array_ptr<const unsigned char> base,
```
Should `base`, `dn`, and `end` have declared bounds?

dopelsunce replied:
`base` and `dn` has `count(0)` by default because they are nt_array_ptrs. The bounds of `end` are not passed as parameters so they cannot be simply specified. 

We may use a bounds cast before the `memcmp` call (line 80) to set bounds `(end - l, end)` to `end`.

kkjeer commented on include/resolv.h:
```diff
@@ -122,22 +122,50 @@ struct res_sym {
+int dn_comp(const char *src,
+	unsigned char *dst : count(space - 1) itype(_Nt_array_ptr<unsigned char>),
+	int space,
+	unsigned char **dnptrs,
```
Should `dnptrs` and `lastdnptr` have bounds-safe interfaces?

dopelsunce committed:
Add comments for dn_comp and associated functions

dopelsunce committed:
Clean up and add comments

dopelsunce commented on include/arpa/nameser.h:
```diff
@@ -345,7 +347,7 @@ int ns_skiprr(const unsigned char *cp : bounds(cp, eom),
 	const unsigned char *eom : itype(_Ptr<const unsigned char>),
 	const unsigned char *src : bounds(src, eom),
-	char *dst : count(dst_size) itype(_Nt_array_ptr<char>),
+	char *dst : count(dst_size) itype(_Array_ptr<char>),
```
The buffer `dst` should use array_ptr because its size is known. Saves the trouble to declare `count(dst_size - 1)` for nt_array_ptr.

dopelsunce commented on include/resolv.h:
```diff
@@ -122,23 +122,51 @@ struct res_sym {
-	const unsigned char *src : bounds(src, end) itype(_Nt_array_ptr<const unsigned char>),
-	char *dest : count(space > 254 ? 254 : space) itype(_Nt_array_ptr<char>),
+	const unsigned char *src : bounds(src, end),
+	char *dest : count(space > 254 ? 254 : space),
```
`src` and `dest` have known bounds and use array_ptr by default.

dopelsunce commented on include/resolv.h:
```diff
@@ -122,23 +122,51 @@ struct res_sym {
 	int space);
 int dn_skipname(const unsigned char *s : bounds(s, end),
-	const unsigned char *end : itype(_Ptr<const unsigned char>));
+	const unsigned char *end : count(0));
```
`end` cannot be dereferenced. It marks the end of `s`.

dopelsunce commented on src/network/dn_comp.c:
```diff
@@ -54,24 +89,36 @@ static int match(int *offset, const unsigned char *base, const unsigned char *dn
+// The limit of the array is specified with lastdnptr. If dnptrs is NULL, domain names are not
+// compressed. If lastdnptr is NULL, the list of labels is not updated.
+int dn_comp(const char *src : itype(_Nt_array_ptr<const char>),
+	unsigned char *dst : count(space - 1) itype(_Nt_array_ptr<unsigned char>),
```
The nt_array_ptr `dst` has `count(space - 1)` because its prefix has at most length `space - 1`. Requires less changes for nt_array_ptr than array_ptr in other use sites.

dopelsunce commented on src/network/dn_comp.c:
```diff
@@ -99,7 +149,8 @@ int dn_comp(const char *src, unsigned char *dst, int space, unsigned char **dnpt
 		if (p+1 < lastdnptr) {
-			*p++ = dst;
+			*p = dst;
+			p++;
```
Using `*p++` causes a compiler error: "inferred bounds 'bounds(*p++, *p++ + 0)' contain a modifying expression; use a temporary instead". Use two lines instead.

dopelsunce commented on src/network/if_nameindex.c:
```diff
@@ -54,7 +54,7 @@ static int netlink_msg_to_nameindex(void *pctx, struct nlmsghdr *h)
 			if (map->index == index &&
 			    map->namelen == namelen &&
-			    memcmp(map->name, RTA_DATA(rta), namelen) == 0)
+			    memcmp((void *)map->name, RTA_DATA(rta), namelen) == 0)
```
`mem*` functions do not accept checked pointers yet.

mgrang replied:
`mem*` functions should now accept checked pointers. See https://github.com/microsoft/checkedc-musl/commit/7d1fec8b8f179b715370c26f17268898ad04cf5a

dopelsunce replied:
Thanks! I removed the cast.

dopelsunce commented on src/network/ns_parse.c:
```diff
@@ -47,15 +47,23 @@ _Checked void ns_put32(unsigned long l, unsigned char *cp : count(4))
+	_Unchecked {
+		handle->_msg = msg;
+		handle->_eom = msg + msglen;
+	}
```
This unchecked block is suggested by David.

dopelsunce commented on src/network/ns_parse.c:
```diff
@@ -168,10 +183,10 @@ int ns_parserr(ns_msg *handle : itype(_Ptr<ns_msg>),
 	const unsigned char *eom : itype(_Ptr<const unsigned char>),
 	const unsigned char *src : bounds(src, eom),
-	char *dst : count(dstsiz) itype(_Nt_array_ptr<char>),
+	char *dst : count(dstsiz) itype(_Array_ptr<char>),
```
`dst` has known size so it should use array_ptr.

dopelsunce committed:
Edit comments

dopelsunce committed:
Merge branch 'master' of https://github.com/microsoft/checkedc-musl into cconv

dopelsunce committed:
Add bounds parameters; remove unchecked blocks

dopelsunce committed:
Pass checked pointers to mem* functions

dopelsunce committed:
Fix typo

kkjeer commented on src/network/dn_comp.c:
```diff
@@ -3,24 +3,45 @@
 		}
+		// s reaches the end of the domain name.
 		if (!*s) return i;
+
```
Please remove the added blank line.

kkjeer commented on src/network/dn_comp.c:
```diff
@@ -54,26 +88,38 @@ static int match(int *offset, const unsigned char *base, const unsigned char *dn
-int dn_comp(const char *src, unsigned char *dst, int space, unsigned char **dnptrs, unsigned char **lastdnptr)
+// dn_comp compresses the domain name src and stores it in the buffer dst of length space. The
+// compresssion uses an array of pointers dnptrs to previously compressed names in the current
+// message. The first pointer points to the beggining of the message and the list ends with NULL.
```
Typo: beggining => beginning

dopelsunce replied:
Thanks

dopelsunce committed:
Use assume_bounds_cast for strings with count(0); fix typo

dopelsunce commented:
I have added a unit test for `dn_comp`: https://github.com/microsoft/checkedc-libc-test/pull/3. The test helped me find several inappropriate bounds declarations, which I have fixed using bounds cast where appropriate.

dopelsunce commented:
Hi @mgrang, would you like to take another look? Thanks!

mgrang reviewed (approved):
LGTM.
